### DISCUSSION TITLE
Auto deploy Preview Deployments in public repos are a security risk

### DISCUSSION THREAD

**Discussion by timomeh** [coollabsio/coolify#3046] [Feature Requests]
The automatic preview deployments in pull requests are a great feature. But right now, in public repos, they seem to be a security risk.

In a public repository, anyone can open a pull request (unless you're on a paid GitHub plan I think, where you can configure who can open a Pull Request). This means that, when preview deployments are enabled, anyone can trigger a new preview deployment and deploy their code by opening a pull request. A malicious actor can access secret environment variables and execute all kinds of shady code. Depending on the domain setup, the domain will also look plausible, it has an SSL certificate, and can be used for social engineering attacks.

- I think this risk should be documented somewhere, at least in the coolify docs.
- It would be great to have a feature to restrict automatically triggered preview deployments to certain users, and (optionally) manually approve them for everyone else. Similar to how you have to approve GitHub Actions to run, or how you have to manually allow a preview deployment with Vercel.

Alternatively to the preview deployments feature, I guess you could use a webhook in a GitHub action, but it's not as good as the Preview Deployment feature.

Please correct me if I'm missing any obvious features that already exist.

**Comment by wdhdev** [comment]
I think it would be good to have some option like `Enable Protected Previews` or something similar which locks the preview deployments behind a login screen so only people that have access to the project can access it.

Currently what I do is I enable automatic preview deployments, but customise deployment hostname for preview deployments to something like `randomid.preview.example.com` and place it behind Cloudflare Access so no one can access those preview deployments except for me.

**Reply by timomeh** [reply]
placing it behind Cloudflare Access is definitely a neat idea! But Protecting who can _access_ the Preview Deployment makes it still vulnerable to security issues, as long as anyone can create a Pull Request which triggers a Preview Deployment. The code can include things that run in the background or on application startup, like grabbing all environment variables and sending them via email or something else.

**Reply by wdhdev** [reply]
Ah yeah, that is true. I didn't think about that!

**Comment by CDE90** [comment]
I would love to see some kind of protection as mentioned here as well. It would be great to be able to use preview deployments in a public repository, but currently I cannot due to security concerns (at least not without setting up custom actions).

**Comment by airtonix** [comment]
best way to deal with this is two steps: 

1. provide a setting in coolify project to opt out of auto preview environments per pull request. 
2. use chat ops github workflow targeting pull requests to create/update/deprovision preview environment for the PR: 
  - coolify instance admins should use this [action](https://github.com/marketplace/actions/slash-command-dispatch) to craft a workflow that only reacts to users in a certain github team.

**Reply by timomeh** [reply]
The [`/deploy` API endpoint](https://coolify.io/docs/api-reference/api/operations/deploy-by-tag-or-uuid) now has a new `pr=` parameter, to trigger a PR preview deployment. This didn't exist back when I opened the discussion.

This means you can now create a GitHub Workflow that triggers the deploy API endpoint, and simply run it on every push to a Pull Requests.

In the GitHub Action settings, you can then configure that workflows need to be approved first for any external user:

<img width="831" alt="image" src="https://github.com/user-attachments/assets/1aee727b-64f3-465d-b5da-640b73c3abd8" />

This effectively gives you an Approve button natively in GitHub's UI to approve a preview deployment, instead of automatically deploying it. You wouldn't even need to craft a custom workflow based on comments.

**Reply by Ayo-Awe** [reply]
The pr parameter doesn't create new preview deployements. It only triggers existing ones
